Allow for multiple installer scripts instead of a singleton

lxea opened the pull request:
#13146  implemented the installer script endpoint with the assumption that it being a singleton would be acceptable however the RFD has been updated to allow for multiple scripts. 

This changes the original implementation so now multiple scripts can be created and looked up by name. It also includes a file-config update so matchers can specify which teleport installer script to use, with a default value of "installer"

```
  aws:
   - types: ["ec2"]
     regions: ["eu-central-1"]
     tags:
       "arch": "ubuntu"
     install:
       script_name: "installer"
```

Commit:
Allow for multiple installer scripts instead of using a singleton

zmb3 commented on lib/web/apiserver.go:
```diff
@@ -459,7 +459,7 @@ func (h *Handler) bindDefaultEndpoints(challengeLimiter *limiter.RateLimiter) {
 	// Unauthenticated access to retrieving the script used to install
 	// Teleport
-	h.GET("/webapi/scripts/installer", httplib.MakeHandler(h.installer))
+	h.GET("/webapi/scripts/:installer", httplib.MakeHandler(h.installer))
```
Would `webapi/scripts/installer/:name` make more sense? What if we need other scripts that are not installers?

Spoiler alert: we are planning on other scripts :-)

Commit:
Resolve comments

- require name be provided to installer
- MetaNameClusterInstallerScript -> defaults.InstallerScriptName
- Require VerbList for GetInstallers

r0mant commented on lib/auth/auth_with_roles.go:
```diff
@@ -3367,11 +3375,19 @@ func (a *ServerWithRoles) SetInstaller(ctx context.Context, inst types.Installer
+
+// DeleteAllInstallers removes all installer script resources
+func (a *ServerWithRoles) DeleteAllInstallers(ctx context.Context) error {
 	if err := a.action(apidefaults.Namespace, types.KindInstaller, types.VerbDelete); err != nil {
```
Include `VerbList` here as well.
```suggestion
	if err := a.action(apidefaults.Namespace, types.KindInstaller, types.VerbDelete, types.VerbList); err != nil {
```

r0mant commented on lib/auth/grpcserver.go:
```diff
@@ -4271,13 +4272,48 @@ func (g *GRPCServer) GetInstaller(ctx context.Context, _ *empty.Empty) (*types.I
 	return inst, nil
 }
 
+func (g *GRPCServer) GetInstallers(ctx context.Context, _ *empty.Empty) (*types.InstallerV1List, error) {
```
```suggestion
// GetInstallers returns all installer script resources registered in the cluster.
func (g *GRPCServer) GetInstallers(ctx context.Context, _ *empty.Empty) (*types.InstallerV1List, error) {
```

r0mant commented on lib/auth/grpcserver.go:
```diff
@@ -4271,13 +4272,48 @@ func (g *GRPCServer) GetInstaller(ctx context.Context, _ *empty.Empty) (*types.I
+		return nil, trace.Wrap(err)
+	}
+	var installersV1 []*types.InstallerV1
+	for _, inst := range res {
```
Should we append the default installer resource here as well?

r0mant commented on lib/auth/grpcserver.go:
```diff
@@ -4252,14 +4253,14 @@ func (g *GRPCServer) SetInstaller(ctx context.Context, req *types.InstallerV1) (
+	res, err := auth.GetInstaller(ctx, req.Name)
 	if err != nil {
-		if trace.IsNotFound(err) {
+		if trace.IsNotFound(err) && req.Name == defaults.InstallerScriptName {
```
This logic will basically allow the users to override the default installer resource if they want to? I think it's a good idea :+1:

r0mant commented on lib/config/fileconf.go:
```diff
@@ -1232,7 +1235,12 @@ type AWSEC2Matcher struct {
+	// JoinParams sets the token and method to use when generating
+	// config on EC2 instances
 	JoinParams JoinParams `yaml:"join_params,omitempty"`
+	// ScriptName is the name of the teleport script for the EC2
```
```suggestion
	// ScriptName is the name of the teleport installer script resource for the EC2
```

r0mant commented on lib/defaults/defaults.go:
```diff
@@ -836,4 +836,7 @@ const (
 	IAMInviteTokenName = "aws-discovery-iam-token"
+	// InstallerScriptName is the name of the by default populated,
+	// EC2 installer script
+	InstallerScriptName = "installer"
```
Should we call it `default-installer` so it's clear that it's a default script?
```suggestion
	InstallerScriptName = "default-installer"
```

r0mant commented on lib/services/configuration.go:
```diff
@@ -71,10 +71,14 @@ type ClusterConfiguration interface {
+	GetInstallers(context.Context) ([]types.Installer, error)
 	// GetInstaller gets the installer script from the backend
-	GetInstaller(context.Context) (types.Installer, error)
+	GetInstaller(context.Context, string) (types.Installer, error)
```
Nit: Use named arguments when it's not clear what the argument is from its type.
```suggestion
	GetInstaller(ctx context.Context, name string) (types.Installer, error)
```

r0mant commented on lib/services/configuration.go:
```diff
@@ -71,10 +71,14 @@ type ClusterConfiguration interface {
 	SetInstaller(context.Context, types.Installer) error
 	// DeleteInstaller removes the installer script from the backend
-	DeleteInstaller(context.Context) error
+	DeleteInstaller(context.Context, string) error
```
```suggestion
	DeleteInstaller(ctx context.Context, name string) error
```

r0mant commented on lib/services/local/configuration.go:
```diff
@@ -401,5 +430,5 @@ const (
 	networkingPrefix       = "networking"
 	sessionRecordingPrefix = "session_recording"
-	installerScriptPrefix  = "installer_script"
+	installerScriptPrefix  = "installer_scripts"
```
Since we're going to be adding other script types soon, how about we store it under `scripts/installer` in the backend? So let's make `scriptsPrefix = "scripts"` and `installerPrefix = "installer"` and assemble the backend key from them.

r0mant replied:
Yes.

r0mant reviewed (approved):
Mostly lgtm, just a few suggestions, please address them before merging.

Commit:
Resolve comments

- Include the default installer in GetInstallers output
- Update how the scripts are prefixed

Commit:
resolve conflicts

Commit:
make the linters happy
